libvast/src/system/connect_to_node.cpp

```diff
@@ -11,28 +11,41 @@
 #include "vast/fwd.hpp"
 
 #include "vast/concept/parseable/vast/endpoint.hpp"
-#include "vast/concept/printable/to_string.hpp"
-#include "vast/concept/printable/vast/port.hpp"
-#include "vast/data.hpp"
 #include "vast/defaults.hpp"
 #include "vast/endpoint.hpp"
 #include "vast/error.hpp"
 #include "vast/logger.hpp"
+#include "vast/system/connector.hpp"
 #include "vast/system/node_control.hpp"
 #include "vast/system/version_command.hpp"
 
-#include <caf/io/middleman.hpp>
-#include <caf/openssl/all.hpp>
 #include <caf/scoped_actor.hpp>
 #include <caf/settings.hpp>
 
 namespace vast::system {
+namespace {
 
-caf::expected<node_actor>
-connect_to_node(caf::scoped_actor& self, const caf::settings& opts) {
-  // Fetch values from config.
-  auto id = get_or(opts, "vast.node-id", defaults::system::node_id.data());
-  auto timeout = node_connection_timeout(opts);
+void check_version(record& remote_version) {
```

Comment:
Should this really take a *non-const* lvalue?

Reply:
I would also argue that the function should return a `bool`.

Reply:
It has to because of the [] operator.
I was thinking about replacing it with .at(), but then i need to also make sure that the key is there and handle the situation where the key is not there. Since this is just a helper function in anonymous namespace i decided to keep it as it was previously.

Also returning boolean would require the caller to handle it. I didn't want to change the behaviour of connect_to_node in this PR.

Reply:
> Also returning boolean would require the caller to handle it. I didn't want to change the behaviour of connect_to_node in this PR.

But you're refactoring the API already, so why stop half way?

> It has to because of the [] operator.

Does anything speak against transforming that into a call to `find`, other than that it's more verbose?

The primary issue is that an API with a signature of `void check_version(record& remote_version)` if not very nice. There should be no need to take non-const reference to express the logging side-effect. The name of the function also suggests that it yields a response. If the version check fails, the you'd want to abort at the call site. (The warnings are no reason to abort, though.) If you wan to keep `operator[]`, you could take a full copy of the input `record`.

More generally, the version check should not be part of the `connector`. Once you have a node actor at the call site, then it's time to the version check there.

Reply:
> The primary issue is that an API with a signature of void check_version(record& remote_version) if not very nice. There should be no need to take non-const reference to express the logging side-effect. The name of the function also suggests that it yields a response. If the version check fails, the you'd want to abort at the call site. (The warnings are no reason to abort, though.) If you wan to keep operator[], you could take a full copy of the input record.

It's not nice but since it's not public (implementation detail in cpp file) i found it to be okayish. I can go with const&, but then what should i do when there is no key available? The current master doesn't care about such cases. Should i return error from connect_to_node? 

> The name of the function also suggests that it yields a response. If the version check fails, the you'd want to abort at the call site. (The warnings are no reason to abort, though.)

If we don't have a good action in case this function fails then i don't see a point of having a return value that is never used. Please notice it's not a public function, it's purpose is to to group logic into smaller more readable block
Maybe i could change the name of the function to be print_versions_mismatch() ?

Reply:
> It's not nice but since it's not public (implementation detail in cpp file) i found it to be okayish. I can go with const&, but then what should i do when there is no key available? 

If there's no key found, that would be a logic error, i.e., assertion, because it violates an invariant. But crashing is not nice, so I'd terminate cleanly. (This is the `false` case that I mention below.)

> The current master doesn't care about such cases. Printing a warning could be fine with the function returning void. Otherwise return false/error and abort the connection procedure?

Yes, return `true` if the check succeeds, and `false` if we cannot continue. 

> Maybe i could change the name of the function to be print_versions_mismatch() ?

No, it's actually a version check. We're just not acting responsibly on it (yet).